core/trino-main/src/main/java/io/trino/server/security/oauth2/NimbusClient.java

```diff
@@ -0,0 +1,445 @@
+                            return null;
+                        }));
+    }
+
+    private void initializeJWTProcessing()
+    {
+        try {
+            jwsKeySelector = new JWSVerificationKeySelector<>(
+                    Stream.concat(JWSAlgorithm.Family.RSA.stream(), JWSAlgorithm.Family.EC.stream()).collect(toImmutableSet()),
+                    new RemoteJWKSet<>(jwksUrl.toURL(), httpClient));
+        }
+        catch (MalformedURLException e) {
+            throw new RuntimeException(e);
+        }
+
+        DefaultJWTProcessor<SecurityContext> processor = new DefaultJWTProcessor<>();
+        processor.setJWSKeySelector(jwsKeySelector);
+        processor.setJWTClaimsSetVerifier(new DefaultJWTClaimsVerifier<>(
+                accessTokenAudiences,
+                new JWTClaimsSet.Builder()
+                        .issuer(accessTokenIssuer.orElse(issuer).getValue())
+                        .build(),
+                ImmutableSet.of(principalField),
+                ImmutableSet.of()));
+        accessTokenProcessor = processor;
+    }
+
+    private <T extends AccessTokenResponse> T getTokenResponse(String code, URI callbackUri, NimbusAirliftHttpClient.Parser<T> parser)
+            throws ChallengeFailedException
+    {
+        T tokenResponse = httpClient.execute(new TokenRequest(tokenUrl, clientAuth, new AuthorizationCodeGrant(new AuthorizationCode(code), callbackUri)), parser);
+        if (!tokenResponse.indicatesSuccess()) {
+            throw new ChallengeFailedException("Error while fetching access token: " + tokenResponse.toErrorResponse().toJSONObject());
+        }
+        return tokenResponse;
+    }
+
+    private Optional<JWTClaimsSet> getJWTClaimsSet(String accessToken)
+    {
+        if (userInfoEndpointEnabled && userinfoUrl.isPresent()) {
+            return queryUserInfo(accessToken);
+        }
+        return parseAccessToken(accessToken);
+    }
+
+    private Optional<JWTClaimsSet> queryUserInfo(String accessToken)
+    {
+        try {
+            UserInfoResponse response = httpClient.execute(new UserInfoRequest(userinfoUrl.get(), new BearerAccessToken(accessToken)), UserInfoResponse::parse);
+            if (!response.indicatesSuccess()) {
+                return Optional.empty();
+            }
+            return Optional.of(response.toSuccessResponse().getUserInfo().toJWTClaimsSet());
+        }
+        catch (ParseException | RuntimeException e) {
+            LOG.error(e, "Received bad response from userinfo endpoint");
+            return Optional.empty();
+        }
+    }
+
+    private Optional<JWTClaimsSet> parseAccessToken(String accessToken)
+    {
+        try {
+            return Optional.of(accessTokenProcessor.process(accessToken, null));
+        }
+        catch (java.text.ParseException | BadJOSEException | JOSEException e) {
+            LOG.error(e, "Failed to parse JWT access token");
+            return Optional.empty();
+        }
+    }
+
+    @VisibleForTesting
```

Comment:
These getters can be package-private or can be removed as AssertJ has a method like `hasFieldOrPropertyWithValue`.